### PR TITLE
Faster implementation for constructing gradient of extract_image_patches

### LINKED ISSUE #20146: tf.extract_image_patches() gradient very slow at graph construction time
### System information
- **Have I written custom code**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Fedora 28
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.8.0-1-g8753e2e
- **Python version**: 3.5
- **Bazel version**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: see below

### Problem
When building the graph, taking the gradient with respect to the input of `tf.extract_image_patches` is unusually slow (see minimal example below, where it takes ~2min). 

### How to reproduce
```python
import tensorflow as tf

tf.reset_default_graph()

batch_size = 4
height = width = 1024
ksize = 5

inputs = tf.get_variable('inputs', (batch_size, height, width, 1))
patches = tf.extract_image_patches(inputs, 
                                   ksizes=[1,ksize,ksize,1], 
                                   strides=[1,1,1,1], 
                                   rates=[1,1,1,1], 
                                   padding='SAME')
print(patches.shape)
%time gradients = tf.gradients(patches, inputs)
```
Result:
```
(4, 1024, 1024, 25)
CPU times: user 1min 42s, sys: 2.07 s, total: 1min 44s
Wall time: 1min 44s
```

### PR DESCRIPTION
**Pull request by facaiy**
Fix #20146.

In the PR, we created indices matrix for input tensor, and reuse `extract_image_patches` to calculate `idx` (the mapping table for indices: input -> output). 

The list below is the performance comparison based on the example of @dbbert in #20146.

Before:
+ constructing time: ~138.4s
+ execute time: ~33.7s

After:
+ constructing time: ~3.6s
+ execute time: ~20.3s

### CODE REVIEW

**Comment by facaiy** [comment]
Test script:

```python
import tensorflow as tf
import time

tf.reset_default_graph()

batch_size = 4
height = width = 1024
ksize = 5

inputs = tf.get_variable('inputs', (batch_size, height, width, 1))
print('inputs: ', inputs)
patches = tf.extract_image_patches(inputs, 
                                   ksizes=[1,ksize,ksize,1], 
                                   strides=[1,1,1,1], 
                                   rates=[1,1,1,1], 
                                   padding='SAME')
print('patches: ', patches.shape)
s = time.clock()
gradients = tf.gradients(patches, inputs)
print('construct time: {}'.format(time.clock() - s))

with tf.Session() as sess:
    print('gradients: ', gradients)
    s = time.clock()
    res = sess.run(gradients)
    print('exec time: {}'.format(time.clock() - s))
    print('res shape: ', res[0].shape)
    print('res: ', res[0][0, 0:3, 0:3, 0])
```

**Comment by facaiy** [comment]
ping @ebrevdo @nikhilmishra000

**Comment by facaiy** [comment]
cc @aselle @suharshs @alextp Could you take a look? Thanks.

**Comment by alextp** [comment]
Can you also add this benchmark to a test, so we can track future regressions?

**Comment by facaiy** [comment]
Sure, could you take a look again? Because execute time is too long (> 20s), we only add benchmark case for constructing gradient here.

File: tensorflow/python/ops/array_grad.py [hunk 1]
```diff
@@ -734,64 +732,59 @@ def _QuantizeAndDequantizeV3Grad(_, grad):
...
+  idx_matrix = array_ops.concat(
+      [array_ops.expand_dims(input_idx_patched, axis=-1),
+       array_ops.expand_dims(output_idx, axis=-1)],
+      axis=-1)
+  idx_map = array_ops.reshape(idx_matrix, (-1, 2))
+
+  sp_shape = (input_indices_num, output_indices_num)
+  sp_mat_full = sparse_tensor.SparseTensor(
+      idx_map,
+      array_ops.ones_like(idx_map[:, 0], dtype=grad.dtype),
```
Comment (facaiy) [inline tensorflow/python/ops/array_grad.py]: I think this line can be replaced by tf.ones(output_indices_num). I'll submit the change later.

**Review by alextp — APPROVED** [review]

**Comment by facaiy** [comment]
@yifeif Hi, yifei. Do you have any update for the PR? Thank you.

**Pull request merged**